doc/design/docker-cp.md

```diff
@@ -0,0 +1,144 @@
+## DOCKER CP DESIGN DOC
+
+### Initial Rundown of docker cp
+
+Docker cp has many possible scenarios that we will have to account for and they will be described in this section.
+
+We will be supporting docker cp in it's entirety. This means we will need to copy files to and from a container regardless of whether it is on or off. This will lead to 4 scenarios. 
+
+1. The container is *on*, and we want to copy data *from* it.
+
+2. The container is *on*, and we want to copy data *to* it.
+
+3. The container is *off*, and we want to copy data *from* it.
+
+4. The container is *off*, an we want to copy data *to* it.
+
+According to the docker remote api the target for a copy must be in `identity (no compression), gzip, bzip2, xz`. As the docker cli handles packaging a *to* request we can expect on of these formats. We must package our *from* response in one of these formats as well. 
+
+ _NOTE_: docker has discussed copying between containers in the past. This will be yet another set of 4 possibilities and endpoints. Something that we can add to this plan later if it is needed in the future. 
+
+### Personality to Portlayer Communication
+
+Since all 4 of these must be supported we will require both distinctions for when the container is off and which endpoint to call to distinguish between *to* and *from* operations. 
+
+The proposed solution is to have two endpoints designed for the portlayer api swagger server. Both endpoints will exist along the same request path as two different verbs. The request path should be as such : 
+
+```
+/container/filesystem?ID=
+```
+
+For a copy *to* the filesystem of a container we will do a `PUT` operation
+
+```
+/contrainer/filesystem?ID= PUT
+
+query param:
+
+* ID - this is the target container
+
+Request Model : 
+
+* path - container filesystem target for the copy operation
+
+Response Model :
+
+* stream - tar stream of the data to be moved to the container
+
+* overwriteDir - flag for whether or not we are directed to overwrite existing directories 
+
+OR 
+
+* Error - error response from the portlayer
+
+```
+
+For a copy *from* the filesystem of a container we will do a `GET` operation. 
+
+```
+/container/filesystem?ID=
+
+Query Param: 
+
+* ID - this is the target container
+
+Request Model : 
+
+* path - target path for the copy operation
+
+* dataStream - tar stream of the data to be copies into the filesystem of the container
+
+OR 
+
+* Error - error response from the portlayer
+
+```
+
+### Personality design
+
+The personality has two endpoints that are called depending on the to/from situation above. Each of those endpoints will call their respective portlayer counterparts. There will be several checks during this process. 
```

Comment:
I'd prefer not to require the caller to know what the current state of the container is - it's a fundamentally racy thing as the state can change between being checked and the copy call being made. Let's have this be a single endpoint and the switching logic done by the portlayer. If you're talking about direction then the phrasing is confusing.

From a practical perspective I suggest thinking about the port layer side as import/export or read/write rather than copyFrom/copyTo. This is because the same paths are needed for `load/import` and `save/export` and copy is actually a subcase of these operations.